### PR TITLE
Fix _backward_norm op registration

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Fix a bug introduced in #9740

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
The updated norm op after #9740 doesn't correctly fallback to the dense implementation when supported options are provided. Also the backward op didn't use the correct param parser for registration, causing fatal messages. 

@anirudh2290 @zheng-da can you help review?

File: tests/python/unittest/test_sparse_ndarray.py [hunk 1]
```diff
@@ -918,18 +918,22 @@ def test_sparse_nd_check_format():
...
-    densities = [0, 0.5]
+    densities = [0, 0.5, 1]
     for stype in stypes:
         for density in densities:
-            check_sparse_nd_norm(stype, shape, density)
+           check_sparse_nd_norm(stype, shape, density, axis=None, keepdims=False, ord=2)
+
+    # test fallback
+    check_sparse_nd_norm(stype, shape, density, axis=0, keepdims=False, ord=2)
+    check_sparse_nd_norm(stype, shape, density, axis=None, keepdims=True, ord=2)
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply tests/python/unittest/test_sparse_ndarray.py]: i had this before but even for dense ord=2 is not implemented

[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply tests/python/unittest/test_sparse_ndarray.py]: Typo: ord != 2

[the remark this answers is not in the document]
Comment (anirudh2290) [reply tests/python/unittest/test_sparse_ndarray.py]: okay. is there a plan to add backward pass test for l2 norm ?

File: src/operator/optimizer_op.cc [hunk 2]
```diff
@@ -305,7 +305,7 @@ inline bool SGDStorageType(const nnvm::NodeAttrs& attrs,
     dispatched = storage_type_assign(out_attrs, static_cast<NDArrayStorageType>(weight_stype),
                                      dispatch_mode, DispatchMode::kFComputeEx);
     // warn users if lazy_update is turned on
-    if (dispatched && param.lazy_update) LogLazyUpdate();
+    if (dispatched && param.wd != 0 && param.lazy_update) LogLazyUpdate();
```
Comment (zheng-da) [inline src/operator/optimizer_op.cc]: is this related?

File: tests/python/unittest/test_sparse_ndarray.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply tests/python/unittest/test_sparse_ndarray.py]: The sparse implementation doesn't support backward and it relies on the automatic fallback mechanism because no FInferStorage is not registered. For these kind of op we don't add extra unit tests. Automatic fallback is tested in  https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_sparse_operator.py#L1619

File: src/operator/tensor/broadcast_reduce_op.h [hunk 3]
```diff
@@ -707,18 +707,16 @@ void ReduceCsr(const nnvm::NodeAttrs& attrs, mshadow::Stream<xpu>* s, const OpCo
...
+                           const std::vector<OpReqType>& req,
+                           const std::vector<NDArray>& outputs) {
   CHECK_EQ(inputs.size(), 1U);
   CHECK_EQ(outputs.size(), 1U);
   CHECK_EQ(req.size(), 1U);
   mshadow::Stream<xpu>* s = ctx.get_stream<xpu>();
   const NDArrayStorageType istype = inputs[0].storage_type();
   if (istype == kCSRStorage) {
-    CHECK_EQ(inputs[0].shape().ndim(), 2U)
-        << "sum(csr)/mean(csr) op only supports 2D ndarray as input";
```
Comment (zheng-da) [inline src/operator/tensor/broadcast_reduce_op.h]: Why this is removed? CSR now supports arrays with more than 2D?

File: src/operator/tensor/broadcast_reduce_op.h [hunk 4]
```diff
@@ -880,17 +878,30 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
...
   if (!dispatched && in_stype == kDefaultStorage) {
     // dns -> dns
     dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
                                      DispatchMode::kFCompute);
   }
-  if (!dispatched && (in_stype == kCSRStorage || in_stype == kRowSparseStorage)) {
-    // csr/rsp -> dns
+  const TShape axis = param.axis.has_value() ? param.axis.value() : TShape();
+  if (!dispatched && (in_stype == kRowSparseStorage || in_stype == kCSRStorage) &&
+      axis.ndim() == 0 && !param.keepdims && param.ord == 2) {
```
Comment (zheng-da) [inline src/operator/tensor/broadcast_reduce_op.h]: does it matter whether keepdims is true?

File: src/operator/optimizer_op.cc (hunk 2 above)
Comment (eric-haibin-lin) [reply src/operator/optimizer_op.cc]: Not really related. But I don't want to make a separate PR for this one line change..

File: src/operator/tensor/broadcast_reduce_op.h (hunk 3 above)
Comment (eric-haibin-lin) [reply src/operator/tensor/broadcast_reduce_op.h]: CSR is always 2D. I don't think a non 2D csr matrix can be constructed, so the check in the operator is not necessary.

File: src/operator/tensor/broadcast_reduce_op.h (hunk 4 above)
Comment (eric-haibin-lin) [reply src/operator/tensor/broadcast_reduce_op.h]: Good catch. I should probably remove this. Actually I found that when axis is None, keepdims has no effect for the dense implementation. 
```

>>> import mxnet as mx
>>> a = mx.nd.ones((2,2))
>>> a.norm()

[ 2.]
<NDArray 1 @cpu(0)>
>>> a.norm(keepdims=True)

[ 2.]
<NDArray 1 @cpu(0)>
```
Is this supposed to return a result of shape (1,1) instead?

File: src/operator/tensor/broadcast_reduce_op.h [hunk 5]
```diff
@@ -880,17 +878,30 @@ inline bool L2NormStorageType(const nnvm::NodeAttrs& attrs,
...
-  if (!dispatched && (in_stype == kCSRStorage || in_stype == kRowSparseStorage)) {
-    // csr/rsp -> dns
+  const TShape axis = param.axis.has_value() ? param.axis.value() : TShape();
+  if (!dispatched && (in_stype == kRowSparseStorage || in_stype == kCSRStorage) &&
+      axis.ndim() == 0 && param.ord == 2) {
+    // l2 norm: rsp/csr, axis = () -> dns
     dispatched = storage_type_assign(&out_stype, kDefaultStorage, dispatch_mode,
                                      DispatchMode::kFComputeEx);
   }
+  if (!dispatched && in_stype == kCSRStorage && axis.ndim() == 1 && !param.keepdims &&
```
Comment (zheng-da) [inline src/operator/tensor/broadcast_reduce_op.h]: the same thing here. checking keepdims here isn't necessary.

Comment (eric-haibin-lin) [reply src/operator/tensor/broadcast_reduce_op.h]: This is because the cpu csr reduce implementation doesn't support keepdims=True.

Comment (eric-haibin-lin) [reply src/operator/tensor/broadcast_reduce_op.h]: If we want to support this, csr.norm(axis=0, keepdims=True) should probably return a csr output instead of dense output
